docs/projects/insertion-sort/_posts/2019-01-22-python.md

```diff
@@ -0,0 +1,218 @@
+
+In this example we convert it into a decimal integer, `int(x.strip(" "), 10)`. Then, `x.strip(" ")`
+removes any whitespace so `" 1"` becomes `"1"`. After that, `int("1", 10)`
+converts the string `"1"` into a decimal number in this case `1`. This is done
+for every item in the list so our original input of `"2, 1, 10, 5, 3"` becomes `[2, 1, 10, 5, 3]`.
+
+### Throw Errors
+
+```python
+def exit_with_error():
+    print('Usage: please provide a list of at least two integers to sort in the format “1, 2, 3, 4, 5”')
+    sys.exit(1)
+```
+
+This function prints a message and then exits the script with an error, `sys.exit(1)`.
+If any non-zero value is returned then the program didn't complete properly. This function is called
+if the user input isn't correct.
+
+### Insertion Sort
+
+```python
+def insertion_sort(xs):
+    new_xs = []
+    for x in xs:
+        new_xs = insert(x, new_xs)
+    return new_xs
+```
+
+Let's take a look at the part of the program that sorts our list. The unsorted list is passed to the
+`insertion_sort` method as the parameter `xs`. Meanwhile, `new_xs = []` is our new sorted listed which is
+empty to begin with.
+
+We loop through every element in the unsorted list `for x in xs`. Then we call the `insert()`
+function to add `x` to the `new_xs` (the sorted list) in the correct position.
+Each time the `insert()` function is called it returns a sorted list which we assign
+to `new_xs`. Finally, when we've looped through every item in `xs`, we return the sorted list
+`new_xs` which will then get printed on the terminal to the user.
+
+Taking a look at an example where `xs = [5, 3, 10]`
+
+1st
+
+* `x = 5`
+* `insert(5, [])`
+* `new_xs = [5]`
+
+2nd
+
+* `x = 3`
+* `insert(3, [5])`
+* `new_xs = [3, 5]`
+
+3rd
+
+* `x = 10`
+* `insert(10, [3, 5])`
+* `new_xs = [3, 5, 10]`
+
+### Insert
+
+```python
+def insert(x, xs):
+    left = list(takewhile(lambda i: i < x, xs))
+    right = xs[len(left):]
+    return left + [x] + right
+```
+
+This function takes two parameters `x`, which is an element from our unsorted list, and
+`xs`, which is the list to add `x` to such that `xs` remains sorted.
+A High-level overview of this function is that the `left` variable will be a list that stores
+all elements less than `x` from `xs` and `right` will store all elements greater than (or equal)
+to `x`. That way we can "insert" `x` between these two lists. Hence the return statement
+looking like `return left + [x] + right` and this returned list will therefore be sorted.
+
+Let's take a look at the first line `left = list(takewhile(lambda i: i < x, xs))` it looks a bit
+complicated so let's break it down. The first part `lambda i: i < x, xs` is a lambda function
+which is a small anonymous unnamed function.
+In this case `i` is an element of `xs` and we want all `i`'s less than `x`.
+
+Then `takewhile(lambda i: i < x, xs)` where `takewhile()` takes a predicate (our lambda function) and a list
```

Comment:
Feel free to reword this as you want. I was trying to find a way to rephrase what was here because it felt a little clunky.

```suggestion
Then, we call `takewhile(lambda i: i < x, xs)` which takes a predicate (our lambda function) and a list.
```